clang-tidy-checks/potential-unbounded-loop-check.diff

```diff
@@ -0,0 +1,3051 @@
+new file mode 100644
+index 000000000000..b00c0525ef4a
+--- /dev/null
++++ b/clang-tools-extra/clang-tidy/misc/PotentialUnboundedLoopCheck.cpp
+@@ -0,0 +1,107 @@
++//===--- PotentialUnboundedLoopCheck.cpp - clang-tidy
++//-------------------------------===//
++//
++// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
++// See https://llvm.org/LICENSE.txt for license information.
++// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
++//
++//===----------------------------------------------------------------------===//
++
++#include "PotentialUnboundedLoopCheck.h"
++#include "clang/AST/ASTContext.h"
++#include "clang/ASTMatchers/ASTMatchFinder.h"
++
++using namespace clang::ast_matchers;
++
++namespace clang {
++namespace tidy {
++namespace misc {
++
++void PotentialUnboundedLoopCheck::registerMatchers(MatchFinder *Finder) {
++  Finder->addMatcher(
++      forStmt(allOf(hasCondition(
++                        binaryOperator(
++                            hasLHS(hasDescendant(declRefExpr().bind("lhs"))),
++                            hasRHS(hasDescendant(declRefExpr().bind("rhs"))),
++                            anyOf(hasOperatorName("<"), hasOperatorName("<="),
++                                  hasOperatorName(">"), hasOperatorName(">=")))
++                            .bind("op")),
++                    hasIncrement(
++                        unaryOperator(
++                            anyOf(hasOperatorName("++"), hasOperatorName("--")),
++                            hasDescendant(declRefExpr().bind("inc")))
++                            .bind("inc_op"))))
++          .bind("for"),
++      this);
++}
++
++inline bool isInc(UnaryOperatorKind Op) {
++  return Op == UnaryOperatorKind::UO_PostInc ||
++         Op == UnaryOperatorKind::UO_PreInc;
++}
++
++inline bool isDec(UnaryOperatorKind Op) {
++  return Op == UnaryOperatorKind::UO_PostDec ||
++         Op == UnaryOperatorKind::UO_PreDec;
++}
++
++inline bool isLT(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_LT;
++}
++
++inline bool isLE(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_LE;
++}
++inline bool isGT(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_GT;
++}
++
++inline bool isGE(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_GE;
++}
++
++void PotentialUnboundedLoopCheck::check(
++    const MatchFinder::MatchResult &Result) {
++  const auto *LHSDecl = Result.Nodes.getNodeAs<DeclRefExpr>("lhs")->getDecl();
++  const auto *RHSDecl = Result.Nodes.getNodeAs<DeclRefExpr>("rhs")->getDecl();
++  const auto *IncDecl = Result.Nodes.getNodeAs<DeclRefExpr>("inc")->getDecl();
++  auto Op = Result.Nodes.getNodeAs<BinaryOperator>("op")->getOpcode();
++  auto IncOp = Result.Nodes.getNodeAs<UnaryOperator>("inc_op")->getOpcode();
++  const auto *MatchedFor = Result.Nodes.getNodeAs<ForStmt>("for");
++
++  // only analyze non-compound types
++  if (LHSDecl->getType()->isCompoundType() ||
```

Comment:
Should we give up for any value thats not an integer type (via `Type::isIntegerType()` instead? For something like this

```cpp
int main() {
    float max = 10000e30;
    for (unsigned long long i = 0; i < max; i++ ) {}
    return 0;
}
```

the widths of `max` and `i` are `32` and `64` respectively, but since it's a float with exponents the actual maximum value it takes on (that's used in the comparison) is much larger than `64` bits. This is a big rabbit hole we probably shouldn't be going down and should just give up early instead